### PR TITLE
feat: upgrade to undici v7

### PR DESCRIPTION
**Pull request by fengmk2**
fix server side close unexpected exit

closes https://github.com/node-modules/urllib/issues/541

## Release Notes

- **New Features**
	- Introduced new example scripts demonstrating HTTP/2 requests with enhanced error handling.
	- Added support for logging response status and memory usage in long-running request scenarios.
	- Added a new `FormData` class to handle file uploads with non-ASCII filenames.
	- Enhanced the `HttpClient` with improved handling of connection closures and multiple concurrent requests.
	- Updated test suite to ensure robust handling of edge cases, including file uploads with special characters.

- **Bug Fixes**
	- Improved error handling in HTTP request processes to better manage socket errors and timeouts.
	- Enhanced diagnostics logging to provide better context for missing data during request processing.

- **Documentation**
	- Updated package configuration to reflect new dependencies and module structure.

- **Refactor**
	- Enhanced type specificity in various components to improve clarity and maintainability.
	- Improved clarity in test cases with better type handling and naming consistency.

### CODE REVIEW

File: examples/longruning.cjs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+});
+
+async function main() {
+  for (let i = 0; i < 1000000; i++) {
+    // await httpClient.request('https://registry.npmmirror.com/');
+    // console.log(r.status, r.headers, r.res.timing);
+    try {
+      const r = await httpClient.request('https://edgeupdates.microsoft.com/api/products');
+      // console.log(r.status, r.headers, r.data.length, r.res.timing);
+      if (i % 10 === 0) {
```
Comment (fengmk2) [inline examples/longruning.cjs:14]: 跑了50000步，rss 没有持续上涨

```bash
50080 403 {
  rss: 107642880,
  heapTotal: 22462464,
  heapUsed: 15521056,
  external: 3602833,
  arrayBuffers: 291214
}
50081 error: other side closed
50090 403 {
  rss: 106381312,
  heapTotal: 19316736,
  heapUsed: 14949512,
  external: 3578815,
  arrayBuffers: 265654
}
50100 403 {
  rss: 105742336,
  heapTotal: 19316736,
  heapUsed: 15079720,
  external: 3591091,
  arrayBuffers: 276460
}
50105 error: other side closed
50108 error: other side closed
50110 403 {
  rss: 106659840,
  heapTotal: 20365312,
  heapUsed: 15311064,
  external: 3595337,
  arrayBuffers: 279236
}
50116 error: other side closed
50120 403 {
  rss: 106364928,
  heapTotal: 19316736,
  heapUsed: 15491176,
  external: 3598034,
  arrayBuffers: 280463
}
50125 error: other side closed
50130 403 {
  rss: 105857024,
  heapTotal: 19316736,
  heapUsed: 15670640,
  external: 3611905,
  arrayBuffers: 292864
}
50140 error: other side closed
50145 error: other side closed
50150 403 {
  rss: 106807296,
  heapTotal: 20365312,
  heapUsed: 16070480,
  external: 3623249,
  arrayBuffers: 300830
}
50159 error: other side closed
50160 403 {
  rss: 107266048,
  heapTotal: 20365312,
  heapUsed: 16151344,
  external: 3628973,
  arrayBuffers: 305012
}
50169 error: other side closed
50170 403 {
  rss: 107298816,
  heapTotal: 20365312,
  heapUsed: 16277176,
  external: 3631871,
  arrayBuffers: 306368
}
50171 error: other side closed
50180 403 {
  rss: 106397696,
  heapTotal: 19316736,
  heapUsed: 16448848,
  external: 3637388,
  arrayBuffers: 310415
}
50181 error: other side closed
50189 error: other side closed
50190 403 {
  rss: 106954752,
  heapTotal: 20365312,
  heapUsed: 12999560,
  external: 3527602,
  arrayBuffers: 238087
}
50194 error: other side closed
50197 error: other side closed
50200 403 {
  rss: 107266048,
  heapTotal: 20365312,
  heapUsed: 12950440,
  external: 3548715,
  arrayBuffers: 257658
}
50201 error: other side closed
50205 error: other side closed
50208 error: other side closed
50210 403 {
  rss: 106364928,
  heapTotal: 19316736,
  heapUsed: 13163288,
  external: 3527654,
  arrayBuffers: 235127
}
50220 403 {
  rss: 105922560,
  heapTotal: 19316736,
  heapUsed: 13151912,
  external: 3535804,
  arrayBuffers: 241807
}
50222 error: other side closed
50226 error: other side closed
50230 403 {
  rss: 105988096,
  heapTotal: 19316736,
  heapUsed: 13413120,
  external: 3566911,
  arrayBuffers: 271225
}
50236 error: other side closed
50240 403 {
  rss: 106856448,
  heapTotal: 20365312,
  heapUsed: 13574016,
  external: 3546076,
  arrayBuffers: 248848
}
50243 error: other side closed
50250 403 {
  rss: 106364928,
  heapTotal: 19316736,
  heapUsed: 13553352,
  external: 3551788,
  arrayBuffers: 253018
}
50260 403 {
  rss: 105709568,
  heapTotal: 19316736,
  heapUsed: 13629192,
  external: 3550710,
  arrayBuffers: 250398
}
50266 error: other side closed
50268 error: other side closed
50270 403 {
  rss: 105791488,
  heapTotal: 19316736,
  heapUsed: 13870400,
  external: 3567282,
  arrayBuffers: 265353
}
50272 error: other side closed
50278 error: other side closed
50280 403 {
  rss: 106708992,
  heapTotal: 20365312,
  heapUsed: 14078880,
  external: 3571400,
  arrayBuffers: 268001
}
50290 403 {
  rss: 106364928,
  heapTotal: 19316736,
  heapUsed: 14241616,
  external: 3575508,
  arrayBuffers: 270639
}
50293 error: other side closed
50300 403 {
  rss: 105873408,
  heapTotal: 19316736,
  heapUsed: 14403472,
  external: 3586561,
  arrayBuffers: 280222
}
50310 403 {
  rss: 105807872,
  heapTotal: 19316736,
  heapUsed: 14488752,
  external: 3585540,
  arrayBuffers: 277587
}
50312 error: other side closed
50320 403 {
  rss: 106676224,
  heapTotal: 20365312,
  heapUsed: 14638496,
  external: 3588587,
  arrayBuffers: 279092
}
50322 error: other side closed
50330 403 {
  rss: 106381312,
  heapTotal: 19316736,
  heapUsed: 14750328,
  external: 3599580,
  arrayBuffers: 288615
}
50335 error: other side closed
50339 error: other side closed
50340 403 {
  rss: 105824256,
  heapTotal: 19316736,
  heapUsed: 14961736,
  external: 3601039,
  arrayBuffers: 288604
}
50344 error: other side closed
50350 403 {
  rss: 105807872,
  heapTotal: 19316736,
  heapUsed: 15095680,
  external: 3609239,
  arrayBuffers: 295334
}
50355 error: other side closed
50360 error: other side closed
50362 error: other side closed
50370 403 {
  rss: 106381312,
  heapTotal: 19316736,
  heapUsed: 15354752,
  external: 3645472,
  arrayBuffers: 328555
}
50373 error: other side closed
50380 error: other side closed
50384 error: other side closed
50390 403 {
  rss: 105791488,
  heapTotal: 19316736,
  heapUsed: 15695096,
  external: 3636945,
  arrayBuffers: 317016
}
50394 error: other side closed
50400 403 {
  rss: 106627072,
  heapTotal: 20365312,
  heapUsed: 15811912,
  external: 3636937,
  arrayBuffers: 315538
}
50403 error: other side closed
50408 error: other side closed
50410 200 {
  rss: 106397696,
  heapTotal: 19316736,
  heapUsed: 15945944,
  external: 3664985,
  arrayBuffers: 342044
}
50415 error: other side closed
50418 error: other side closed
50420 403 {
  rss: 106397696,
  heapTotal: 19316736,
  heapUsed: 16615504,
  external: 3722511,
  arrayBuffers: 398028
}
50424 error: other side closed
50430 403 {
  rss: 106348544,
  heapTotal: 19316736,
  heapUsed: 16796512,
  external: 3677986,
  arrayBuffers: 352033
}
50434 error: other side closed
50440 403 {
  rss: 106381312,
  heapTotal: 20365312,
  heapUsed: 12659496,
  external: 3523570,
  arrayBuffers: 234643
}
50444 error: other side closed
50450 403 {
  rss: 106020864,
  heapTotal: 19316736,
  heapUsed: 12713568,
  external: 3530497,
  arrayBuffers: 240100
}
50454 error: other side closed
50460 403 {
  rss: 105807872,
  heapTotal: 19316736,
  heapUsed: 12798552,
  external: 3530778,
  arrayBuffers: 238911
}
50463 error: other side closed
50469 error: other side closed
50470 403 {
  rss: 105873408,
  heapTotal: 19316736,
  heapUsed: 13022744,
  external: 3534884,
  arrayBuffers: 241547
}
50472 error: other side closed
50477 error: other side closed
50480 403 {
  rss: 105922560,
  heapTotal: 19316736,
  heapUsed: 13219656,
  external: 3549864,
  arrayBuffers: 255057
}
50488 error: other side closed
50490 403 {
  rss: 106840064,
  heapTotal: 20365312,
  heapUsed: 13243784,
  external: 3548657,
  arrayBuffers: 252308
}
50498 error: other side closed
50500 403 {
  rss: 107413504,
  heapTotal: 20365312,
  heapUsed: 13489648,
  external: 3552927,
  arrayBuffers: 255108
}
50510 403 {
  rss: 107413504,
  heapTotal: 20365312,
  heapUsed: 13570472,
  external: 3562534,
  arrayBuffers: 263245
}
50513 error: other side closed
50520 403 {
  rss: 107413504,
  heapTotal: 20365312,
  heapUsed: 13805776,
  external: 3568064,
  arrayBuffers: 267305
}
50527 error: other side closed
50530 403 {
  rss: 107544576,
  heapTotal: 22462464,
  heapUsed: 13944752,
  external: 3579289,
  arrayBuffers: 276988
}
50531 error: other side closed
50540 403 {
  rss: 107741184,
  heapTotal: 22462464,
  heapUsed: 14733472,
  external: 3606405,
  arrayBuffers: 302634
}
50542 error: other side closed
50550 403 {
  rss: 106545152,
  heapTotal: 19316736,
  heapUsed: 14220256,
  external: 3582141,
  arrayBuffers: 276900
}
50553 error: other side closed
50557 error: other side closed
50560 403 {
  rss: 105807872,
  heapTotal: 19316736,
  heapUsed: 14389720,
  external: 3621075,
  arrayBuffers: 314292
}
50570 error: other side closed
50580 error: other side closed
50586 error: other side closed
50590 403 {
  rss: 105758720,
  heapTotal: 19316736,
  heapUsed: 14815624,
  external: 3608656,
  arrayBuffers: 297463
}
50598 error: other side closed
50600 403 {
  rss: 106594304,
  heapTotal: 20365312,
  heapUsed: 14958472,
  external: 3611493,
  arrayBuffers: 298830
}
50606 error: other side closed
50610 403 {
  rss: 107380736,
  heapTotal: 22462464,
  heapUsed: 15115848,
  external: 3614377,
  arrayBuffers: 300172
}
50611 error: other side closed
50620 403 {
  rss: 107462656,
  heapTotal: 22462464,
  heapUsed: 15887160,
  external: 3649672,
  arrayBuffers: 333997
}
50628 error: other side closed
50630 403 {
  rss: 108167168,
  heapTotal: 22462464,
  heapUsed: 16651432,
  external: 3676626,
  arrayBuffers: 359481
}
50633 error: other side closed
50638 error: other side closed
50640 403 {
  rss: 108363776,
  heapTotal: 22462464,
  heapUsed: 16230968,
  external: 3651275,
  arrayBuffers: 332660
}
50645 error: other side closed
50649 error: other side closed
50650 403 {
  rss: 109068288,
  heapTotal: 22462464,
  heapUsed: 12427672,
  external: 3504404,
  arrayBuffers: 216065
}
50660 403 {
  rss: 109068288,
  heapTotal: 22462464,
  heapUsed: 13140792,
  external: 3532607,
  arrayBuffers: 242798
}
50670 403 {
  rss: 109068288,
  heapTotal: 22462464,
  heapUsed: 13893640,
  external: 3560786,
  arrayBuffers: 269507
}
50671 error: other side closed
50680 200 {
  rss: 106545152,
  heapTotal: 19316736,
  heapUsed: 13332600,
  external: 3559711,
  arrayBuffers: 266890
}
50684 error: other side closed
50690 403 {
  rss: 107249664,
  heapTotal: 20365312,
  heapUsed: 13497408,
  external: 3537328,
  arrayBuffers: 243037
}
50694 error: other side closed
50700 403 {
  rss: 106545152,
  heapTotal: 19316736,
  heapUsed: 13605512,
  external: 3539019,
  arrayBuffers: 243258
}
50702 error: other side closed
50708 error: other side closed
50710 403 {
  rss: 106004480,
  heapTotal: 19316736,
  heapUsed: 13603920,
  external: 3545904,
  arrayBuffers: 248673
}
50711 error: other side closed
50716 error: other side closed
50720 403 {
  rss: 105971712,
  heapTotal: 19316736,
  heapUsed: 13743784,
  external: 3550071,
  arrayBuffers: 251298
}
50721 error: other side closed
50725 error: other side closed
50730 403 {
  rss: 106905600,
  heapTotal: 20365312,
  heapUsed: 13926488,
  external: 3556904,
  arrayBuffers: 256661
}
50732 error: other side closed
50738 error: other side closed
50740 403 {
  rss: 106414080,
  heapTotal: 19316736,
  heapUsed: 14153832,
  external: 3569229,
  arrayBuffers: 267516
}
50747 error: other side closed
50750 403 {
  rss: 106119168,
  heapTotal: 19316736,
  heapUsed: 14363880,
  external: 3600194,
  arrayBuffers: 296792
}
50753 error: other side closed
50760 403 {
  rss: 105971712,
  heapTotal: 19316736,
  heapUsed: 14358824,
  external: 3575003,
  arrayBuffers: 270131
}
50766 error: other side closed
50770 403 {
  rss: 106889216,
  heapTotal: 20365312,
  heapUsed: 14592416,
  external: 3584792,
  arrayBuffers: 278450
}
50775 error: other side closed
50780 403 {
  rss: 106364928,
  heapTotal: 19316736,
  heapUsed: 14762640,
  external: 3590435,
  arrayBuffers: 282623
}
50785 error: other side closed
50790 403 {
  rss: 106004480,
  heapTotal: 19316736,
  heapUsed: 14894000,
  external: 3617112,
  arrayBuffers: 307758
}
```

File: src/FormData.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import path from 'node:path';
+import _FormData from 'form-data';
+
+export class FormData extends _FormData {
+  _getContentDisposition(value: any, options: any) {
+    // support non-ascii filename
+    // https://github.com/form-data/form-data/pull/571
```
Comment (fengmk2) [inline src/FormData.ts:7]: https://github.com/form-data/form-data/pull/571

File: src/HttpClient.ts [hunk 3]
```diff
@@ -612,6 +604,19 @@ export class HttpClient extends EventEmitter {
         res.size = parseInt(res.headers['content-length']);
       }
 
+      // https://developer.mozilla.org/en-US/docs/Web/HTTP/Redirections
+      if (RedirectStatusCodes.includes(res.statusCode) && maxRedirects > 0 && requestContext.redirects < maxRedirects && !isStreamingRequest) {
```
Comment (fengmk2) [inline src/HttpClient.ts:608]: 最终还是靠自己解决

**Comment by fengmk2** [comment]
先发一个 beta 版本在 cnpmcore 验证完没问题，再合并发布一个新版本

**Comment by fengmk2** [comment]
> 先发一个 beta 版本在 cnpmcore 验证完没问题，再合并发布一个新版本

https://github.com/cnpm/cnpmcore/pull/734

**Comment by fengmk2** [comment]
@coderabbitai 提一个 issue 将 https://github.com/form-data/form-data/pull/571 修复 pick 到 urllib v3 分支
